src/gfx/convert.cpp

```diff
@@ -0,0 +1,831 @@
+                           std::vector<Palette> const &palettes,
+                           DefaultInitVec<size_t> const &mappings) {
+	std::filebuf output;
+	output.open(options.output, std::ios_base::out | std::ios_base::binary);
+
+	auto iter = attrmap.begin();
+	for (auto tile : png.visitAsTiles(options.columnMajor)) {
+		Palette const &palette = palettes[mappings[iter->protoPaletteID]];
+		for (uint32_t y = 0; y < 8; ++y) {
+			uint16_t row = 0;
+			for (uint32_t x = 0; x < 8; ++x) {
+				row <<= 1;
+				uint8_t index = palette.indexOf(tile.pixel(x, y));
+				if (index & 1) {
+					row |= 0x001;
+				}
+				if (index & 2) {
+					row |= 0x100;
+				}
+			}
+			output.sputc(row & 0xFF);
+			output.sputc(row >> 8);
+		}
+		++iter;
+	}
+	assert(iter == attrmap.end());
+}
+
+static void outputMaps(Png const &png, DefaultInitVec<AttrmapEntry> const &attrmap,
+                       DefaultInitVec<size_t> const &mappings) {
+	std::optional<std::filebuf> tilemapOutput, attrmapOutput;
+	if (!options.tilemap.empty()) {
+		tilemapOutput.emplace();
+		tilemapOutput->open(options.tilemap, std::ios_base::out | std::ios_base::binary);
+	}
+	if (!options.attrmap.empty()) {
+		attrmapOutput.emplace();
+		attrmapOutput->open(options.attrmap, std::ios_base::out | std::ios_base::binary);
+	}
+
+	uint8_t tileID = 0;
+	uint8_t bank = 0;
+	auto iter = attrmap.begin();
+	for ([[maybe_unused]] auto tile : png.visitAsTiles(options.columnMajor)) {
+		if (tileID == options.maxNbTiles[bank]) {
+			assert(bank == 0);
+			bank = 1;
+			tileID = 0;
+		}
+
+		if (tilemapOutput.has_value()) {
+			tilemapOutput->sputc(tileID + options.baseTileIDs[bank]);
+		}
+		if (attrmapOutput.has_value()) {
+			uint8_t palID = mappings[iter->protoPaletteID] & 7;
+			attrmapOutput->sputc(palID | bank << 3); // The other flags are all 0
+			++iter;
+		}
+		++tileID;
+	}
+}
+
+} // namespace unoptimized
+
+static uint8_t flip(uint8_t byte) {
+	// To flip all the bits, we'll flip both nibbles, then each nibble half, etc.
+	byte = (byte & 0x0F) << 4 | (byte & 0xF0) >> 4;
+	byte = (byte & 0x33) << 2 | (byte & 0xCC) >> 2;
+	byte = (byte & 0x55) << 1 | (byte & 0xAA) >> 1;
+	return byte;
+}
+
+class TileData {
+	// TODO: might want to switch to `std::byte` instead?
+	std::array<uint8_t, 16> _data;
+	// The hash is a bit lax: it's the XOR of all lines, and every other nibble is identical
+	// if horizontal mirroring is in effect. It should still be a reasonable tie-breaker in
+	// non-pathological cases.
+	uint16_t _hash;
+public:
+	mutable size_t tileID;
+
+	TileData(Png::TilesVisitor::Tile const &tile, Palette const &palette) : _hash(0) {
+		for (uint32_t y = 0; y < 8; ++y) {
+			uint16_t bitplanes = 0;
+			for (uint32_t x = 0; x < 8; ++x) {
+				bitplanes <<= 1;
+				uint8_t index = palette.indexOf(tile.pixel(x, y));
+				if (index & 1) {
+					bitplanes |= 1;
+				}
+				if (index & 2) {
+					bitplanes |= 0x100;
+				}
+			}
```

Comment:
This is identical to the code in another file except it's called `bitplanes` instead of `row`. How about factoring it out into a utility function or method?

```c++
uint16_t get_bitplane(palette, tile, y) {
	uint16_t row = 0;
	for (uint32_t x = 0; x < 8; ++x) {
		row <<= 1;
		uint8_t index = palette.indexOf(tile.pixel(x, y));
		if (index & 1) {
			row |= 1;
		}
		if (index & 2) {
			row |= 0b1'0000'0000;
		}
	}
	return row;
}
```

Reply:
I don't disagree.

https://github.com/ISSOtm/rgbds/blob/75d54f439492a960ae5930aa26797789cd244ffe/src/gfx/convert.cpp#L424